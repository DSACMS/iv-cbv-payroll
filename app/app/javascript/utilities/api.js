import { fetchInternal } from "./fetchInternal"
import CSRF from "./csrf"

export const EVENTS_USER_ACTION = "/api/events/user_action"
export const PINWHEEL_TOKENS_GENERATE = "/api/pinwheel/tokens"
const ARGYLE_TOKENS_GENERATE = "/api/argyle/tokens"

export const trackUserAction = async (eventName, attributes = {}) => {
  return fetchInternal(EVENTS_USER_ACTION, {
    method: "post",
    body: JSON.stringify({ events: { event_name: eventName, attributes } }),
  })
}

export const fetchPinwheelToken = async (response_type, id, locale) => {
  const response = await fetch(PINWHEEL_TOKENS_GENERATE, {
    method: "post",
    headers: {
      "X-CSRF-Token": CSRF.token,
      "Content-Type": "application/json",
    },
    // Keep payload shape consistent with existing backend handling
    body: JSON.stringify({ response_type, id, locale }),
  })

  if (response.redirected) {
    window.location.href = response.url
    return
  }

  return response.json()
}

export const fetchArgyleToken = async (itemId) => {
  const response = await fetch(ARGYLE_TOKENS_GENERATE, {
    method: "post",
    headers: {
      "X-CSRF-Token": CSRF.token,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ item_id: itemId }),
  })

  if (response.redirected) {
    window.location.href = response.url
    return
  }

  return response.json()
}
