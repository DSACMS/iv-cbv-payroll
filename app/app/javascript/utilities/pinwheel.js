import loadScript from 'load-script';
import metaContent from "./meta";
import CSRF from './csrf';

export function loadPinwheel() {
  return new Promise((resolve, reject) => {
    loadScript('https://cdn.getpinwheel.com/pinwheel-v3.0.js', (err, script) => {
      if (err) {
        reject(err);
      } else {
        resolve(Pinwheel);
      }
    });
  });
}

export function initializePinwheel(Pinwheel, linkToken, callbacks) {
  return Pinwheel.open({
    linkToken,
    sandbox: metaContent('pinwheel_sandbox'), // Set to false for production environment.
    ...callbacks
  });
}
