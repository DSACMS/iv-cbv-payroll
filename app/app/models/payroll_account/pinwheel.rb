class PayrollAccount::Pinwheel < PayrollAccount
  JOBS_TO_WEBHOOK_EVENTS = {
    # Mapping of job name (in supported jobs) to the webhook event name that
    # signifies its completion.
    "paystubs" => "paystubs.fully_synced",
    "employment" => "employment.added",
    "income" => "income.added",
    "identity" => "identity.added",
    "shifts" => "shifts.added"
  }

  def has_fully_synced?
    (supported_jobs.exclude?("paystubs") || find_webhook_event(JOBS_TO_WEBHOOK_EVENTS["paystubs"]).present?) &&
      (supported_jobs.exclude?("shifts") || find_webhook_event(JOBS_TO_WEBHOOK_EVENTS["shifts"]).present?) &&
      (supported_jobs.exclude?("employment") || find_webhook_event(JOBS_TO_WEBHOOK_EVENTS["employment"]).present?) &&
      (supported_jobs.exclude?("income") || find_webhook_event(JOBS_TO_WEBHOOK_EVENTS["income"]).present?) &&
      (supported_jobs.exclude?("identity") || find_webhook_event(JOBS_TO_WEBHOOK_EVENTS["identity"]).present?)
  end

  def job_succeeded?(job)
    job_status(job) == :succeeded
  end

  def job_status(job)
    if supported_jobs.exclude?(job)
      :unsupported
    elsif find_webhook_event(JOBS_TO_WEBHOOK_EVENTS[job], :success).present?
      :succeeded
    elsif find_webhook_event(JOBS_TO_WEBHOOK_EVENTS[job], :error).present?
      :failed
    else
      :in_progress
    end
  end

  def necessary_jobs_succeeded?
    job_succeeded?("paystubs") || (job_succeeded?("employment") && job_succeeded?("identity"))
  end

  def redact!
    # Do nothing, as Pinwheel does not support the deletion of data via API.
    touch(:redacted_at)
  end
end
