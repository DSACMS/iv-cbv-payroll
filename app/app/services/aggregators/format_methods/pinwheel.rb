module Aggregators::FormatMethods::Pinwheel
  # See on Google Drive:
  # "Pinwheel Payroll Providers (Sandbox/Production) 2024-07-02"
  # "Mapping in Code" sheet
  #
  # We have to do this by matching employer_name because there is no employer
  # ID in any of the pinwheel endpoints.
  GIG_PLATFORM_NAMES = [
    "Airbnb (Host)", "Amazon Flex", "Bite Squad", "Care.com", "DoorDash (Dasher)",
    "Ebay (Seller)", "Etsy", "Field Agent", "GrubHub (Driver)", "Handy",
    "Instacart (Full Service Shopper)", "Lyft (Driver)", "OnlyFans",
    "Patreon (Freelancer)", "Poshmark (Seller)", "Postmates (Fleet)",
    "Roadie", "Shipt (Shopper)", "Shopify Store", "Thumbtack, Inc.",
    "Twitch", "Uber (Driver)", "Via (Driver)", "Wonolo"
  ]

  def self.hours(earnings)
    base_hours = earnings
      .filter { |e| e["category"] != "overtime" }
      .map { |e| e["hours"] }
      .compact
      .max
    return unless base_hours

    # Add overtime hours to the base hours, because they tend to be additional
    # work beyond the other entries. (As opposed to category="premium", which
    # often duplicates other earnings' hours.)
    #
    # See FFS-1773.
    overtime_hours = earnings
      .filter { |e| e["category"] == "overtime" }
      .sum { |e| e["hours"] || 0.0 }

    base_hours + overtime_hours
  end

  def self.hours_by_earning_category(earnings)
    earnings
      .filter { |e| e["hours"] && e["hours"] > 0 }
      .group_by { |e| e["category"] }
      .transform_values { |earnings| earnings.sum { |e| e["hours"] } }
  end

  def self.employment_type(employer_name)
    if GIG_PLATFORM_NAMES.include?(employer_name)
      :gig
    else
      :w2
    end
  end
end
