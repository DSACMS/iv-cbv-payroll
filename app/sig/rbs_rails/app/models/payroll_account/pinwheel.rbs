class PayrollAccount < ::ApplicationRecord
  class Pinwheel < ::PayrollAccount
    extend ::_ActiveRecord_Relation_ClassMethods[::PayrollAccount::Pinwheel, ::PayrollAccount::Pinwheel::ActiveRecord_Relation, ::Integer]

    module GeneratedAttributeMethods
      def id: () -> ::Integer

      def id=: (::Integer) -> ::Integer

      def id?: () -> bool

      def id_changed?: () -> bool

      def id_change: () -> [ ::Integer?, ::Integer? ]

      def id_will_change!: () -> void

      def id_was: () -> ::Integer?

      def id_previously_changed?: () -> bool

      def id_previous_change: () -> ::Array[::Integer?]?

      def id_previously_was: () -> ::Integer?

      def id_before_last_save: () -> ::Integer?

      def id_change_to_be_saved: () -> ::Array[::Integer?]?

      def id_in_database: () -> ::Integer?

      def saved_change_to_id: () -> ::Array[::Integer?]?

      def saved_change_to_id?: () -> bool

      def will_save_change_to_id?: () -> bool

      def restore_id!: () -> void

      def clear_id_change: () -> void

      def cbv_flow_id: () -> ::Integer

      def cbv_flow_id=: (::Integer) -> ::Integer

      def cbv_flow_id?: () -> bool

      def cbv_flow_id_changed?: () -> bool

      def cbv_flow_id_change: () -> [ ::Integer?, ::Integer? ]

      def cbv_flow_id_will_change!: () -> void

      def cbv_flow_id_was: () -> ::Integer?

      def cbv_flow_id_previously_changed?: () -> bool

      def cbv_flow_id_previous_change: () -> ::Array[::Integer?]?

      def cbv_flow_id_previously_was: () -> ::Integer?

      def cbv_flow_id_before_last_save: () -> ::Integer?

      def cbv_flow_id_change_to_be_saved: () -> ::Array[::Integer?]?

      def cbv_flow_id_in_database: () -> ::Integer?

      def saved_change_to_cbv_flow_id: () -> ::Array[::Integer?]?

      def saved_change_to_cbv_flow_id?: () -> bool

      def will_save_change_to_cbv_flow_id?: () -> bool

      def restore_cbv_flow_id!: () -> void

      def clear_cbv_flow_id_change: () -> void

      def income_synced_at: () -> ::ActiveSupport::TimeWithZone?

      def income_synced_at=: (::ActiveSupport::TimeWithZone?) -> ::ActiveSupport::TimeWithZone?

      def income_synced_at?: () -> bool

      def income_synced_at_changed?: () -> bool

      def income_synced_at_change: () -> [ ::ActiveSupport::TimeWithZone?, ::ActiveSupport::TimeWithZone? ]

      def income_synced_at_will_change!: () -> void

      def income_synced_at_was: () -> ::ActiveSupport::TimeWithZone?

      def income_synced_at_previously_changed?: () -> bool

      def income_synced_at_previous_change: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def income_synced_at_previously_was: () -> ::ActiveSupport::TimeWithZone?

      def income_synced_at_before_last_save: () -> ::ActiveSupport::TimeWithZone?

      def income_synced_at_change_to_be_saved: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def income_synced_at_in_database: () -> ::ActiveSupport::TimeWithZone?

      def saved_change_to_income_synced_at: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def saved_change_to_income_synced_at?: () -> bool

      def will_save_change_to_income_synced_at?: () -> bool

      def restore_income_synced_at!: () -> void

      def clear_income_synced_at_change: () -> void

      def created_at: () -> ::ActiveSupport::TimeWithZone

      def created_at=: (::ActiveSupport::TimeWithZone) -> ::ActiveSupport::TimeWithZone

      def created_at?: () -> bool

      def created_at_changed?: () -> bool

      def created_at_change: () -> [ ::ActiveSupport::TimeWithZone?, ::ActiveSupport::TimeWithZone? ]

      def created_at_will_change!: () -> void

      def created_at_was: () -> ::ActiveSupport::TimeWithZone?

      def created_at_previously_changed?: () -> bool

      def created_at_previous_change: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def created_at_previously_was: () -> ::ActiveSupport::TimeWithZone?

      def created_at_before_last_save: () -> ::ActiveSupport::TimeWithZone?

      def created_at_change_to_be_saved: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def created_at_in_database: () -> ::ActiveSupport::TimeWithZone?

      def saved_change_to_created_at: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def saved_change_to_created_at?: () -> bool

      def will_save_change_to_created_at?: () -> bool

      def restore_created_at!: () -> void

      def clear_created_at_change: () -> void

      def updated_at: () -> ::ActiveSupport::TimeWithZone

      def updated_at=: (::ActiveSupport::TimeWithZone) -> ::ActiveSupport::TimeWithZone

      def updated_at?: () -> bool

      def updated_at_changed?: () -> bool

      def updated_at_change: () -> [ ::ActiveSupport::TimeWithZone?, ::ActiveSupport::TimeWithZone? ]

      def updated_at_will_change!: () -> void

      def updated_at_was: () -> ::ActiveSupport::TimeWithZone?

      def updated_at_previously_changed?: () -> bool

      def updated_at_previous_change: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def updated_at_previously_was: () -> ::ActiveSupport::TimeWithZone?

      def updated_at_before_last_save: () -> ::ActiveSupport::TimeWithZone?

      def updated_at_change_to_be_saved: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def updated_at_in_database: () -> ::ActiveSupport::TimeWithZone?

      def saved_change_to_updated_at: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def saved_change_to_updated_at?: () -> bool

      def will_save_change_to_updated_at?: () -> bool

      def restore_updated_at!: () -> void

      def clear_updated_at_change: () -> void

      def supported_jobs: () -> ::String?

      def supported_jobs=: (::String?) -> ::String?

      def supported_jobs?: () -> bool

      def supported_jobs_changed?: () -> bool

      def supported_jobs_change: () -> [ ::String?, ::String? ]

      def supported_jobs_will_change!: () -> void

      def supported_jobs_was: () -> ::String?

      def supported_jobs_previously_changed?: () -> bool

      def supported_jobs_previous_change: () -> ::Array[::String?]?

      def supported_jobs_previously_was: () -> ::String?

      def supported_jobs_before_last_save: () -> ::String?

      def supported_jobs_change_to_be_saved: () -> ::Array[::String?]?

      def supported_jobs_in_database: () -> ::String?

      def saved_change_to_supported_jobs: () -> ::Array[::String?]?

      def saved_change_to_supported_jobs?: () -> bool

      def will_save_change_to_supported_jobs?: () -> bool

      def restore_supported_jobs!: () -> void

      def clear_supported_jobs_change: () -> void

      def type: () -> ::String

      def type=: (::String) -> ::String

      def type?: () -> bool

      def type_changed?: () -> bool

      def type_change: () -> [ ::String?, ::String? ]

      def type_will_change!: () -> void

      def type_was: () -> ::String?

      def type_previously_changed?: () -> bool

      def type_previous_change: () -> ::Array[::String?]?

      def type_previously_was: () -> ::String?

      def type_before_last_save: () -> ::String?

      def type_change_to_be_saved: () -> ::Array[::String?]?

      def type_in_database: () -> ::String?

      def saved_change_to_type: () -> ::Array[::String?]?

      def saved_change_to_type?: () -> bool

      def will_save_change_to_type?: () -> bool

      def restore_type!: () -> void

      def clear_type_change: () -> void

      def synchronization_status: () -> ::String?

      def synchronization_status=: (::String?) -> ::String?

      def synchronization_status?: () -> bool

      def synchronization_status_changed?: () -> bool

      def synchronization_status_change: () -> [ ::String?, ::String? ]

      def synchronization_status_will_change!: () -> void

      def synchronization_status_was: () -> ::String?

      def synchronization_status_previously_changed?: () -> bool

      def synchronization_status_previous_change: () -> ::Array[::String?]?

      def synchronization_status_previously_was: () -> ::String?

      def synchronization_status_before_last_save: () -> ::String?

      def synchronization_status_change_to_be_saved: () -> ::Array[::String?]?

      def synchronization_status_in_database: () -> ::String?

      def saved_change_to_synchronization_status: () -> ::Array[::String?]?

      def saved_change_to_synchronization_status?: () -> bool

      def will_save_change_to_synchronization_status?: () -> bool

      def restore_synchronization_status!: () -> void

      def clear_synchronization_status_change: () -> void

      def redacted_at: () -> ::ActiveSupport::TimeWithZone?

      def redacted_at=: (::ActiveSupport::TimeWithZone?) -> ::ActiveSupport::TimeWithZone?

      def redacted_at?: () -> bool

      def redacted_at_changed?: () -> bool

      def redacted_at_change: () -> [ ::ActiveSupport::TimeWithZone?, ::ActiveSupport::TimeWithZone? ]

      def redacted_at_will_change!: () -> void

      def redacted_at_was: () -> ::ActiveSupport::TimeWithZone?

      def redacted_at_previously_changed?: () -> bool

      def redacted_at_previous_change: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def redacted_at_previously_was: () -> ::ActiveSupport::TimeWithZone?

      def redacted_at_before_last_save: () -> ::ActiveSupport::TimeWithZone?

      def redacted_at_change_to_be_saved: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def redacted_at_in_database: () -> ::ActiveSupport::TimeWithZone?

      def saved_change_to_redacted_at: () -> ::Array[::ActiveSupport::TimeWithZone?]?

      def saved_change_to_redacted_at?: () -> bool

      def will_save_change_to_redacted_at?: () -> bool

      def restore_redacted_at!: () -> void

      def clear_redacted_at_change: () -> void

      def aggregator_account_id: () -> ::String?

      def aggregator_account_id=: (::String?) -> ::String?

      def aggregator_account_id?: () -> bool

      def aggregator_account_id_changed?: () -> bool

      def aggregator_account_id_change: () -> [ ::String?, ::String? ]

      def aggregator_account_id_will_change!: () -> void

      def aggregator_account_id_was: () -> ::String?

      def aggregator_account_id_previously_changed?: () -> bool

      def aggregator_account_id_previous_change: () -> ::Array[::String?]?

      def aggregator_account_id_previously_was: () -> ::String?

      def aggregator_account_id_before_last_save: () -> ::String?

      def aggregator_account_id_change_to_be_saved: () -> ::Array[::String?]?

      def aggregator_account_id_in_database: () -> ::String?

      def saved_change_to_aggregator_account_id: () -> ::Array[::String?]?

      def saved_change_to_aggregator_account_id?: () -> bool

      def will_save_change_to_aggregator_account_id?: () -> bool

      def restore_aggregator_account_id!: () -> void

      def clear_aggregator_account_id_change: () -> void
    end
    include GeneratedAttributeMethods
    def webhook_events: () -> ::WebhookEvent::ActiveRecord_Associations_CollectionProxy
    def webhook_events=: (::WebhookEvent::ActiveRecord_Associations_CollectionProxy | ::Array[::WebhookEvent]) -> (::WebhookEvent::ActiveRecord_Associations_CollectionProxy | ::Array[::WebhookEvent])
    def webhook_event_ids: () -> ::Array[::Integer]
    def webhook_event_ids=: (::Array[::Integer]) -> ::Array[::Integer]

    def cbv_flow: () -> ::CbvFlow
    def cbv_flow=: (::CbvFlow?) -> ::CbvFlow?
    def reload_cbv_flow: () -> ::CbvFlow?
    def build_cbv_flow: (untyped) -> ::CbvFlow
    def create_cbv_flow: (untyped) -> ::CbvFlow
    def create_cbv_flow!: (untyped) -> ::CbvFlow

    module GeneratedRelationMethods
    end

    class ActiveRecord_Relation < ::ActiveRecord::Relation
      include ::PayrollAccount::Pinwheel::GeneratedRelationMethods
      include ::_ActiveRecord_Relation[::PayrollAccount::Pinwheel, ::Integer]
      include ::Enumerable[::PayrollAccount::Pinwheel]
    end

    class ActiveRecord_Associations_CollectionProxy < ::ActiveRecord::Associations::CollectionProxy
      include ::PayrollAccount::Pinwheel::GeneratedRelationMethods
      include ::_ActiveRecord_Relation[::PayrollAccount::Pinwheel, ::Integer]
    end
  end
end
