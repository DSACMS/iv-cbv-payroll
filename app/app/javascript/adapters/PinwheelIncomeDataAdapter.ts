import { fetchToken, trackUserAction } from "../utilities/api";
import loadScript from 'load-script';
import { getDocumentLocale } from "../utilities/getDocumentLocale";
import { IncomeDataAdapter } from "./IncomeDataAdapter";

declare global {
    var Pinwheel: any;
}

export default class PinwheelIncomeDataAdapter extends IncomeDataAdapter {
    Pinwheel: any;

    async load() {
        this.Pinwheel = await new Promise((resolve, reject) => {
            loadScript('https://cdn.getpinwheel.com/pinwheel-v3.0.js', (err, script) => {
                if (err) {
                    reject(err);
                } else {
                    resolve(Pinwheel);
                }
            });
        })
    }

    async open() {
        const locale = getDocumentLocale();
        await trackUserAction("ApplicantSelectedEmployerOrPlatformItem", {
            item_type: this.requestData.responseType,
            item_id: this.requestData.id,
            item_name: this.requestData.name,
            is_default_option: this.requestData.isDefaultOption,
            provider_name: this.requestData.providerName,
            locale
        })

        const { token } = await fetchToken(this.requestData.responseType, this.requestData.id, locale);

        return this.Pinwheel.open({
            linkToken: token,
            onSuccess: this.onSuccess.bind(this),
            onExit: this.onExit.bind(this),
            onEvent: this.onEvent.bind(this),
        });
    }


    async onSuccess(eventPayload: {
        accountId: string;
        platformId: string;
    }) {
        await trackUserAction("PinwheelSuccess", {
            account_id: eventPayload.accountId,
            platform_id: eventPayload.platformId
        })
        if (this.successCallback) {
            this.successCallback(eventPayload.accountId);
        }
    }

    onEvent(eventName: string, eventPayload: any) {
        switch (eventName) {
            case "screen_transition":
                onScreenTransitionEvent(eventPayload.screenName);
                break;
            case 'login_attempt':
                trackUserAction("PinwheelAttemptLogin", {})
                break;
            case 'error':
                const { type, code, message } = eventPayload
                trackUserAction("PinwheelError", { type, code, message })
                break;
            case 'exit':
                trackUserAction("PinwheelCloseModal", {})
                break;
                
        }


        function onScreenTransitionEvent(screenName) {
            switch (screenName) {
                case "LOGIN":
                    trackUserAction("PinwheelShowLoginPage", {
                        screen_name: screenName,
                        employer_name: eventPayload.selectedEmployerName,
                        platform_name: eventPayload.selectedPlatformName
                    });
                    break;
                case "PROVIDER_CONFIRMATION":
                    trackUserAction("PinwheelShowProviderConfirmationPage", {});
                    break;
                case "SEARCH_DEFAULT":
                    trackUserAction("PinwheelShowDefaultProviderSearch", {});
                    break;
                case "EXIT_CONFIRMATION":
                    trackUserAction("PinwheelAttemptClose", {});
                    break;
            }
        }
    }
}