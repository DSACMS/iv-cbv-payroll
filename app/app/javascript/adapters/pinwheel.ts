import { fetchToken, trackUserAction } from "../utilities/api";
import loadScript from 'load-script';
import { getDocumentLocale } from "../utilities/getDocumentLocale";

declare global {
    var Pinwheel: any;
}

export const createProvider = (providerName: string) => {
    switch(providerName) {
        default:
            return PinwheelIncomeDataAdapter;
    }
}
abstract class IncomeDataAdapter {
    successCallback?: Function;
    exitCallback?: Function;
    
    abstract open(responseType : string, id : string, name : string, isDefaultOption : boolean): void;
    abstract onEvent(eventName : string, eventPayload : any): void;
    abstract load();

    constructor(args : {
        onSuccess?: Function;
        onExit?: Function;
    } = { onSuccess: () => {}}) {
        this.successCallback = args.onSuccess;
        this.exitCallback = args.onExit;

        this.load();
    }

    async onExit(eventPayload: any) {
        if (this.exitCallback) {
            this.exitCallback()
        }
    }

    async onSuccess(eventPayload: any) {
        if (this.successCallback) {
            this.successCallback()
        }
    }
}

export default class PinwheelIncomeDataAdapter extends IncomeDataAdapter {
    Pinwheel: any;

    async load() {
        this.Pinwheel = await new Promise((resolve, reject) => {
            loadScript('https://cdn.getpinwheel.com/pinwheel-v3.0.js', (err, script) => {
                if (err) {
                    reject(err);
                } else {
                    resolve(Pinwheel);
                }
            });
        })
    }

    async open(responseType, id, name, isDefaultOption) {
        const locale = getDocumentLocale();
        await trackUserAction("ApplicantSelectedEmployerOrPlatformItem", {
            item_type: responseType,
            item_id: id,
            item_name: name,
            is_default_option: isDefaultOption,
            locale
        })

        const { token } = await fetchToken(responseType, id, locale);

        return this.Pinwheel.open({
            linkToken: token,
            onSuccess: this.onSuccess.bind(this),
            onExit: this.onExit.bind(this),
            onEvent: this.onEvent.bind(this),
        });
    }


    async onSuccess(eventPayload: {
        accountId: string;
        platformId: string;
    }) {
        await trackUserAction("PinwheelSuccess", {
            account_id: eventPayload.accountId,
            platform_id: eventPayload.platformId
        })
        if (this.successCallback) {
            this.successCallback(eventPayload.accountId);
        }
    }

    onEvent(eventName: string, eventPayload: any) {
        switch (eventName) {
            case "screen_transition":
                onScreenTransitionEvent(eventPayload.screenName);
                break;
            case 'login_attempt':
                trackUserAction("PinwheelAttemptLogin", {})
                break;
            case 'error':
                const { type, code, message } = eventPayload
                trackUserAction("PinwheelError", { type, code, message })
                break;
            case 'exit':
                trackUserAction("PinwheelCloseModal", {})
                break;
                
        }


        function onScreenTransitionEvent(screenName) {
            switch (screenName) {
                case "LOGIN":
                    trackUserAction("PinwheelShowLoginPage", {
                        screen_name: screenName,
                        employer_name: eventPayload.selectedEmployerName,
                        platform_name: eventPayload.selectedPlatformName
                    });
                    break;
                case "PROVIDER_CONFIRMATION":
                    trackUserAction("PinwheelShowProviderConfirmationPage", {});
                    break;
                case "SEARCH_DEFAULT":
                    trackUserAction("PinwheelShowDefaultProviderSearch", {});
                    break;
                case "EXIT_CONFIRMATION":
                    trackUserAction("PinwheelAttemptClose", {});
                    break;
            }
        }
    }
}