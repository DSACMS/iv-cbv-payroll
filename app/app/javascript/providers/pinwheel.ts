import { initializePinwheel } from "../utilities/pinwheel";
import { fetchToken, trackUserAction } from "../utilities/api";
import loadScript from 'load-script';
import { getDocumentLocale } from "../utilities/getDocumentLocale";

declare global {
    var Pinwheel: any;
}

export const createProvider = (providerName: string) => {
    if (providerName == "pinwheel") {
        return PinwheelProviderWrapper;
    }
}
abstract class ProviderWrapper {
    abstract successCallback?: Function;
    abstract exitCallback?: Function;
    
    abstract open(responseType : string, id : string, name : string, isDefaultOption : boolean): void;
    abstract onEvent(eventName : string, eventPayload : any): void;
}

export default class PinwheelProviderWrapper extends ProviderWrapper {
    Pinwheel: any;
    successCallback?: Function;
    exitCallback?: Function;

    constructor(args : {
        onSuccess?: Function;
        onExit?: Function;
    } = { onSuccess: () => {}}) {
        super();
        this.successCallback = args.onSuccess;
        this.exitCallback = args.onExit;

        this.load();
    }

    async load() {
        this.Pinwheel = await new Promise((resolve, reject) => {
            loadScript('https://cdn.getpinwheel.com/pinwheel-v3.0.js', (err, script) => {
                if (err) {
                    reject(err);
                } else {
                    resolve(Pinwheel);
                }
            });
        })
    }

    async open(responseType, id, name, isDefaultOption) {
        const locale = getDocumentLocale();
        await trackUserAction("ApplicantSelectedEmployerOrPlatformItem", {
            item_type: responseType,
            item_id: id,
            item_name: name,
            is_default_option: isDefaultOption,
            locale
        })

        const { token } = await fetchToken(responseType, id, locale);

        return this.Pinwheel.open({
            linkToken: token,
            onSuccess: this.onSuccess.bind(this),
            onExit: this.onExit.bind(this),
            onEvent: this.onEvent.bind(this),
        });
    }

    async onExit() {
        if (this.exitCallback) {
            this.exitCallback()
        }
    }
    async onSuccess(eventPayload: {
        accountId: string;
        platformId: string;
    }) {
        await trackUserAction("PinwheelSuccess", {
            account_id: eventPayload.accountId,
            platform_id: eventPayload.platformId
        })
        if (this.successCallback) {
            this.successCallback(eventPayload.accountId);
        }
    }

    onEvent(eventName: string, eventPayload: any) {
        if (eventName === 'screen_transition') {
            onScreenTransitionEvent();
        } else if (eventName === 'login_attempt') {
            trackUserAction("PinwheelAttemptLogin", {})
        } else if (eventName === 'error') {
            const { type, code, message } = eventPayload
            trackUserAction("PinwheelError", { type, code, message })
        } else if (eventName === 'exit') {
            trackUserAction("PinwheelCloseModal", {})
        }


        function onScreenTransitionEvent() {
            const { screenName } = eventPayload;

            switch (screenName) {
                case "LOGIN":
                    trackUserAction("PinwheelShowLoginPage", {
                        screen_name: screenName,
                        employer_name: eventPayload.selectedEmployerName,
                        platform_name: eventPayload.selectedPlatformName
                    });
                    break;
                case "PROVIDER_CONFIRMATION":
                    trackUserAction("PinwheelShowProviderConfirmationPage", {});
                    break;
                case "SEARCH_DEFAULT":
                    trackUserAction("PinwheelShowDefaultProviderSearch", {});
                    break;
                case "EXIT_CONFIRMATION":
                    trackUserAction("PinwheelAttemptClose", {});
                    break;
            }
        }
    }
}