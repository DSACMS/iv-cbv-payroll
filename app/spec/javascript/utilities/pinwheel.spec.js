import { vi, describe, beforeEach, afterEach, it, expect } from "vitest";
import loadScript from "load-script";
import PinwheelIncomeDataAdapter from "@js/providers/pinwheel";
import { fetchToken, trackUserAction } from '@js/utilities/api';
import { mockPinwheel } from "@test/fixtures/pinwheel.fixture";

const mockPinwheelAuthToken = { token: 'test-token' };
const MOCK_PINWHEEL_ERROR = "Failed to load SCRIPT"

const pinwheelIncomeDataAdapterArgs = {
    onSuccess: vi.fn(),
}



describe('PinwheelAdapter', () => {
    let PinwheelAdapter;
    let triggers;
        
    beforeEach(async () => {
        mockPinwheel();
        PinwheelAdapter = new PinwheelIncomeDataAdapter(pinwheelIncomeDataAdapterArgs)
        triggers = await PinwheelAdapter.open("response-type", "id", "name", false)
    })
    afterEach(() => {
    })

    describe('open', () => {
        it('should call track user action', async () => {
            expect(trackUserAction).toHaveBeenCalled()
            expect(trackUserAction.mock.calls[0][0]).toBe('ApplicantSelectedEmployerOrPlatformItem')
            expect(trackUserAction.mock.calls[0]).toMatchSnapshot()
        })
        it('should fetch token successfully', async () => {
            expect(fetchToken).toHaveBeenCalledTimes(1)
            expect(fetchToken).toHaveBeenCalledWith("response-type", "id", "en")
            expect(fetchToken).toHaveResolvedWith({ token: 'test-token' })
        })
        it('should open Pinwheel modal', async () => {
            expect(Pinwheel.open).toHaveBeenCalledTimes(1)
        })
    })

    describe('onSuccess', () => {
        it('should call track user action', async() => {
            await triggers.triggerSuccessEvent()
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelSuccess')
        })
        it('should trigger the provided onSuccess callback', async () => {
            await triggers.triggerSuccessEvent()
            expect(pinwheelIncomeDataAdapterArgs.onSuccess).toHaveBeenCalled()
        })
    })
    describe('onExit', () => {
        it('should trigger the provided onExit callback', async () => {
            await triggers.triggerSuccessEvent()
            expect(pinwheelIncomeDataAdapterArgs.onSuccess).toHaveBeenCalled()
        })
    })
    describe('onEvent', () => {
        it('should log screen_transition.LOGIN Event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "LOGIN", selectedEmployerName: "ACME Inc", selectedPlatformName: "ADP" })
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowLoginPage')
        })
        it('should log screen_transition.PROVIDER_CONFIRMATION event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "PROVIDER_CONFIRMATION"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowProviderConfirmationPage')
        })
        it('should log screen_transition.SEARCH_DEFAULT event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "SEARCH_DEFAULT"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowDefaultProviderSearch')
        })
        it('should log screen_transition.EXIT_CONFIRMATION event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "EXIT_CONFIRMATION"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelAttemptClose')
        })
        it('should log login_attempt event', async () => {
            await triggers.triggerEvent("login_attempt")
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelAttemptLogin')
        })
        it('should log error event', async () => {
            await triggers.triggerEvent("error", { type: "error-type", "code": "code", "message":"default message"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelError')
            expect(trackUserAction.mock.calls[1][1]['type']).toBe('error-type')
        })
        it('should log exit event', async () => {
            await triggers.triggerEvent("exit")
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelCloseModal')
        })
    })
})