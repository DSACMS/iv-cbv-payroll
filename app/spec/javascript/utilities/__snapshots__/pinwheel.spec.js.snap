// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`initializePinwheel > opens Pinwheel modal 1`] = `
[
  {
    "linkToken": "link-token",
    "onError": [MockFunction spy],
    "onSuccess": [MockFunction spy],
  },
]
`;
